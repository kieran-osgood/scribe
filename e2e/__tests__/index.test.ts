import spawnAsync from '@expo/spawn-async';
import path from 'path';
import stripAnsi from 'strip-ansi';

import { arrowKey, createMinimalProject } from '../fixtures';

const cliPath = path.join(process.cwd(), 'dist', 'index.js');
const configFlag = path.join('scribe.config.ts');

describe('Scribe Cli', () => {
  describe('--help', function () {
    it('should print help text', async () => {
      const t = await spawnAsync(cliPath, [`--help`]);

      expect(stripAnsi(t.stdout)).toMatchInlineSnapshot(`
        "━━━ scribe ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
          $ scribe <command>

        ━━━ General commands ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

          scribe [-c,--config #0] [--verbose] [-n,--name #0] [-t,--template #0]
            Scribe generates files based on mustache templates.

          scribe init [-c,--config #0] [--verbose]
            Generates a scribe.config.ts file.

        You can also print more details about any of these commands by calling them with 
        the \`-h,--help\` flag right after the command name.
        "
      `);

      expect(t.status).toBe(0);
    });
  });

  describe('--config & fully interactive', () => {
    it('should complete successfully', async () => {
      const projectRoot = createMinimalProject();

      const processPromise = spawnAsync(
        cliPath,
        [
          `--config=${configFlag}`, //
        ],
        { cwd: projectRoot },
      );

      const cli = processPromise.child;

      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
      if (cli === null) {
        throw new Error('spawned process is null');
      }
      // cli.stderr?.pipe(process.stderr);

      cli.stdout?.on('data', (data: { toString(): string }) => {
        if (data.toString().includes('Pick your template')) {
          cli.stdin?.write(`${arrowKey.down}\n`);
        } else if (data.toString().includes('File name')) {
          cli.stdin?.write('Login\n');
        } else {
          // console.log(data.toString());
        }
      });
      // .pipe(process.stdout);

      const result = await processPromise;

      const stdOutAnsi = stripAnsi(result.stdout);
      expect(stdOutAnsi).toMatch(
        '✅  Success!\n' +
          'Output files:\n' +
          `- ${projectRoot}/examples/src/screens/Login.ts\n` +
          `- ${projectRoot}/examples/src/screens/Login.test.ts\n`,
      );
      expect(result.status).toBe(0);
    });

    describe('--template --name', () => {
      it('should complete successfully', async () => {
        const projectRoot = createMinimalProject();

        const result = await spawnAsync(
          cliPath,
          [`--config=${configFlag}`, '--template=screen', '--name=Login'],
          { cwd: projectRoot },
        );

        const stdOutAnsi = stripAnsi(result.stdout);
        expect(stdOutAnsi).toMatch(
          '✅  Success!\n' +
            'Output files:\n' +
            `- ${projectRoot}/examples/src/screens/Login.ts\n` +
            `- ${projectRoot}/examples/src/screens/Login.test.ts\n`,
        );
        expect(result.status).toBe(0);
      });
    });
  });
});
