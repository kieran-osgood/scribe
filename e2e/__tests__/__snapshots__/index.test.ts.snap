// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DefaultCommand > [Given] Flag passed in: --config & fully interactive session > [Then] creates two files 1`] = `
"// @ts-ignore
import * as React from 'react';

type LoginProps = {}
function LoginScreen() {

}
"
`;

exports[`DefaultCommand > [Given] Flag passed in: --config & fully interactive session > [Then] creates two files 2`] = `
"describe('Login', function() {
  it('should ', function() {

  });
});
"
`;

exports[`DefaultCommand > [Given] all flags used: --config --template --name > [Then] creates two files 1`] = `
"// @ts-ignore
import * as React from 'react';

type LoginProps = {}
function LoginScreen() {

}
"
`;

exports[`DefaultCommand > [Given] all flags used: --config --template --name > [Then] creates two files 2`] = `
"describe('Login', function() {
  it('should ', function() {

  });
});
"
`;

exports[`DefaultCommand > [Given] only --name passed in > [Then] creates two files 1`] = `
"// @ts-ignore
import * as React from 'react';

type LoginProps = {}
function LoginScreen() {

}
"
`;

exports[`DefaultCommand > [Given] only --name passed in > [Then] creates two files 2`] = `
"describe('Login', function() {
  it('should ', function() {

  });
});
"
`;

exports[`DefaultCommand > [Given] only --template passed in > [Then] creates two files 1`] = `
"// @ts-ignore
import * as React from 'react';

type LoginProps = {}
function LoginScreen() {

}
"
`;

exports[`DefaultCommand > [Given] only --template passed in > [Then] creates two files 2`] = `
"describe('Login', function() {
  it('should ', function() {

  });
});
"
`;

exports[`InitCommand > [Given] Git Clean > [When] filepath clear [Then] create file 2`] = `
"import { ScribeConfig } from '@kieran-osgood/scribe';

export default {
  options: {
    rootOutDir: '.',
    templatesDirectories: ['.'],
  },
  templates: {},
} satisfies ScribeConfig;
"
`;

exports[`InitCommand > [Given] Git Clean > [When] filepath clear [Then] create file 3`] = `
"import { ScribeConfig } from '@kieran-osgood/scribe';

export default {
  options: {
    rootOutDir: '.',
    templatesDirectories: ['.'],
  },
  templates: {},
} satisfies ScribeConfig;
"
`;

exports[`InitCommand > [Given] Git Clean > [When] filepath full [then] print failure 2`] = `
"import type { ScribeConfig } from '@scribe/config';

const config = {
  options: {
    rootOutDir: '.',
    templatesDirectories: ['./src/common/test-fixtures'],
  },
  templates: {
    component: {
      outputs: [
        {
          templateFileKey: 'component',
          output: {
            directory: 'examples/src/components',
            fileName: '{{Name}}.ts',
          },
        },
      ],
    },
    screen: {
      outputs: [
        {
          templateFileKey: 'screen',
          output: {
            directory: 'examples/src/screens',
            fileName: '{{Name}}.ts',
          },
        },
        {
          templateFileKey: 'screen.test',
          output: {
            directory: 'examples/src/screens',
            fileName: '{{Name}}.test.ts',
          },
        },
      ],
    },
  },
} satisfies ScribeConfig;

export default config;
"
`;

exports[`InitCommand > [Given] Git dirty > [Then] prompt user to continue > [When] user accepts [Then] create file 2`] = `
"import { ScribeConfig } from '@kieran-osgood/scribe';

export default {
  options: {
    rootOutDir: '.',
    templatesDirectories: ['.'],
  },
  templates: {},
} satisfies ScribeConfig;
"
`;

exports[`InitCommand > [Given] Git dirty > [Then] prompt user to continue > [When] user accepts [Then] create file 3`] = `
"import { ScribeConfig } from '@kieran-osgood/scribe';

export default {
  options: {
    rootOutDir: '.',
    templatesDirectories: ['.'],
  },
  templates: {},
} satisfies ScribeConfig;
"
`;
